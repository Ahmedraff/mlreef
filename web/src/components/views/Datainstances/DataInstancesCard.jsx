import React from 'react';
import { toastr } from 'react-redux-toastr';
import DataInstanteDeleteModal from 'components/DeleteDataInstance/DeleteDatainstance';
import {
  CANCELED, FAILED, PENDING, RUNNING, SUCCESS,
} from 'dataTypes';
import { Link } from 'router';
import { getPipelineIcon, getInfoFromStatus } from 'functions/pipeLinesHelpers';
import DataInstanceActions from './DataInstanceActions';

const DataInstancesCard = (props) => {
  const {
    params,
    history,
    namespace,
    slug,
    fetchPipelines,
    fireModal,
  } = props;
  
  const { statusTitle } = getInfoFromStatus(params?.currentState);

  function getButtonsDiv(instance) {
    let buttons;
    const { currentState } = instance;
    if (currentState === RUNNING || currentState === PENDING) {
      buttons = [
        <button
          type="button"
          key="abort-button"
          className="btn btn-danger border-solid my-auto"
          style={{ width: 'max-content' }}
          onClick={() => {
            fireModal({
              title: `Abort ${instance.descTitle}`,
              type: 'danger',
              closable: true,
              content: <DataInstanteDeleteModal dataInstanceName={instance.descTitle} />,
              onPositive: () => DataInstanceActions.abortDataInstance(
                instance.projId,
                instance.pipelineBackendId,
                instance.backendInstanceId,
                instance.id,
              )
                .then(fetchPipelines)
                .then(() => toastr.success('Success', 'The data instace was aborted'))
                .catch((err) => toastr.error('Error', err?.message)),
            });
          }}
        >
          Abort
        </button>,
      ];
    } else if (
      currentState === SUCCESS
    ) {
      buttons = [
        <button
          type="button"
          key="experiment-button"
          className="btn btn-outline-dark my-auto mr-1"
          onClick={() => history.push(`/${namespace}/${slug}/-/datasets/${instance?.pipelineBackendId}/rebuild`)}
        >
          View Pipeline
        </button>,
        <button
          type="button"
          key="delete-button"
          onClick={
              () => {
                fireModal({
                  title: `Delete ${instance.descTitle}`,
                  type: 'danger',
                  closable: true,
                  content: <DataInstanteDeleteModal dataInstanceName={instance.descTitle} />,
                  onPositive: () => {
                    DataInstanceActions.deleteDataInstance(
                      instance.pipelineBackendId,
                      instance.backendInstanceId,
                    )
                      .then(fetchPipelines)
                      .then(() => toastr.success('Success', 'Pipeline was deleted'))
                      .catch((error) => toastr.error('Error', error?.message));
                  },
                });
              }
            }
          className="btn btn-danger btn-icon my-auto"
        >
          <i className="fa fa-times" />
        </button>,
      ];
    } else if (currentState === FAILED
        || currentState === CANCELED) {
      buttons = [
        <button
          type="button"
          key="experiment-button"
          className="btn btn-outline-dark my-auto mr-1"
          onClick={() => history.push(`/${namespace}/${slug}/-/datasets/${instance?.pipelineBackendId}/rebuild`)}
        >
          View Pipeline
        </button>,
        <button
          type="button"
          key="delete-button"
          className="btn btn-danger btn-icon my-auto"
          onClick={() => {
            fireModal({
              title: `Delete ${instance.descTitle}`,
              type: 'danger',
              closable: true,
              content: <DataInstanteDeleteModal dataInstanceName={instance.descTitle} />,
              onPositive: () => {
                DataInstanceActions.deleteDataInstance(
                  instance.pipelineBackendId,
                  instance.backendInstanceId,
                )
                  .then(fetchPipelines)
                  .then(() => toastr.success('Success', 'Pipeline was deleted'))
                  .catch((error) => toastr.error('Error', error?.message));
              },
            });
          }}
        >
          <i className="fa fa-times" />
        </button>,
      ];
    }

    return (
      <div className="buttons-div d-flex">{buttons}</div>
    );
  }
  return (
    <div className="pipeline-card">
      <div className="header">
        <div className="title-div">
          <p><b>{statusTitle}</b></p>
        </div>
      </div>

      {params?.instances?.map((instance) => {
        const {
          id: dataId, pipelineBackendId, userName, currentState,
        } = instance;
        const dataInstanceName = instance.descTitle;
        const uniqueName = dataInstanceName.split('/')[1];
        const modelDiv = 'inherit';
        let progressVisibility = 'inherit';
        if (currentState === 'Expired') { progressVisibility = 'hidden'; }
        return (
          <div key={`instance-comp-id-${instance.id}`} className="card-content">
            <div id="data-ins-summary-data" className="summary-data" data-key={`${instance.descTitle}`}>
              <img style={{ alignSelf: 'center' }} src={getPipelineIcon(currentState)} width="30" height="30" alt={currentState} />
              <div className="project-desc-experiment">
                <Link to={`/${namespace}/${slug}/-/datasets/${pipelineBackendId}`}>
                  <b>{uniqueName}</b>
                </Link>
                <p className="m-0 mt-1">
                  Created by
                  {' '}
                  <a href={`/${userName}`}>
                    <b>
                      {userName}
                    </b>
                  </a>
                  <br />
                  {instance.timeCreatedAgo}
                  {' '}
                  ago
                </p>
              </div>
              <div className="project-desc-experiment" style={{ visibility: progressVisibility }}>
                <p><b>Usage: ---</b></p>
              </div>
              <div className="project-desc-experiment d-flex" style={{ visibility: modelDiv }}>
                <p style={{ flex: '1' }}>
                  Id:
                  {' '}
                  {dataId}
                </p>
                { getButtonsDiv(instance) }
              </div>
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default DataInstancesCard;
