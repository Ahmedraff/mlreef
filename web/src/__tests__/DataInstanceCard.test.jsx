import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { shallow } from 'enzyme';
import DataInstancesCard from 'components/views/Datainstances/DataInstancesCard';

const props = {
  name: 'data-pipeline/dear-nessie-4112020-1',
  namespace: 'mlreef',
  slug: 'mlreef-slug',
  currentState: 'failed',
  instances: [
    {
      backendInstanceId: 'be307d64-3237-49f8-898b-46831f99aabc',
      backendProjectId: '5d005488-afb6-4a0c-852a-f471153a04b5',
      commitId: '57ce6ced1e790ee0a164680a9be46aeba33f1316',
      currentState: 'failed',
      descTitle: 'data-pipeline/dear-nessie-4112020-1',
      id: 16,
      pipelineBackendId: 'e3397d9c-6f0f-4607-b41b-fe27720ff8ac',
      projId: 20,
      timeCreatedAgo: '4 hour(s)',
      userName: 'mlreef',
    },
  ],
  fireModal: jest.fn(),
};

test('assert that snapshot matches', () => {
  const snapShot = renderer.create(
    <Router>
      <DataInstancesCard
        key={props.name}
        name={props.name}
        namespace={props.namespace}
        slug={props.slug}
        params={{
          currentState: props.currentState,
          instances: props.instances,
        }}
        fireModal={props.fireModal}
      />
    </Router>,
  ).toJSON();
  expect(snapShot).toMatchSnapshot();
});

const setup = () => {
  const testWrapper = shallow(
    <DataInstancesCard
      key={props.name}
      name={props.name}
      namespace={props.namespace}
      slug={props.slug}
      params={{
        currentState: props.currentState,
        instances: props.instances,
      }}
      fireModal={props.fireModal}
    />,
  );

  return testWrapper;
};

describe('assert basic UI text elements are rendered', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = setup();
  });
  test('pipeline info is displayed', () => {
    expect(wrapper.find('img').prop('alt')).toEqual('failed');
    expect(wrapper.find('.title-div').text()).toBe('Failed');
    expect(wrapper.find('Link').children().text()).toBe('dear-nessie-4112020-1');
    expect(wrapper.find('a').text()).toBe('mlreef');
  });

  test('assert that pipeline functinal buttons are rendered', () => {
    expect(wrapper.find('.buttons-div').children()).toHaveLength(2);
  });
});
