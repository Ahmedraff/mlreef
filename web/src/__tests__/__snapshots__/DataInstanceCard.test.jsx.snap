// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assert that snapshot matches 1`] = `
<div
  className="pipeline-card"
>
  <div
    className="header"
  >
    <div
      className="title-div"
    >
      <p>
        <b>
          Failed
        </b>
      </p>
    </div>
  </div>
  <div
    className="card-content"
  >
    <div
      className="summary-data"
      data-key="data-pipeline/dear-nessie-4112020-1"
      id="data-ins-summary-data"
    >
      <img
        alt="failed"
        height="30"
        src={
          Object {
            "canInstrument": true,
            "createTransformer": [Function],
            "getCacheKey": [Function],
            "process": [Function],
          }
        }
        style={
          Object {
            "alignSelf": "center",
          }
        }
        width="30"
      />
      <div
        className="project-desc-experiment"
      >
        <a
          className=""
          href="/mlreef/mlreef-slug/-/datasets/e3397d9c-6f0f-4607-b41b-fe27720ff8ac"
          onClick={[Function]}
        >
          <b>
            dear-nessie-4112020-1
          </b>
        </a>
        <p
          className="m-0 mt-1"
        >
          Created by
           
          <a
            href="/mlreef"
          >
            <b>
              mlreef
            </b>
          </a>
          <br />
          4 hour(s)
           
          ago
        </p>
      </div>
      <div
        className="project-desc-experiment"
        style={
          Object {
            "visibility": "inherit",
          }
        }
      >
        <p>
          <b>
            Usage: ---
          </b>
        </p>
      </div>
      <div
        className="project-desc-experiment d-flex"
        style={
          Object {
            "visibility": "inherit",
          }
        }
      >
        <p
          style={
            Object {
              "flex": "1",
            }
          }
        >
          Id:
           
          16
        </p>
        <div
          className="buttons-div d-flex"
        >
          <button
            className="btn btn-outline-dark my-auto mr-1"
            onClick={[Function]}
            type="button"
          >
            View Pipeline
          </button>
          <button
            className="btn btn-danger btn-icon my-auto"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-times"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
